Index: BaseTools/Conf/tools_def.template
===================================================================
--- BaseTools/Conf/tools_def.template	(revision 1)
+++ BaseTools/Conf/tools_def.template	(working copy)
@@ -15,13 +15,13 @@
 IDENTIFIER = Default TOOL_CHAIN_CONF
 
 # common path macros
-DEFINE VS2003_BIN       = C:\Program Files\Microsoft Visual Studio .NET 2003\Vc7\bin
-DEFINE VS2003_DLL       = C:\Program Files\Microsoft Visual Studio .NET 2003\Common7\IDE
+DEFINE VS2003_BIN       = ENV(VS2003_PATH)\Vc7\bin
+DEFINE VS2003_DLL       = ENV(VS2003_PATH)\Common7\IDE
 
-DEFINE VS2005_BIN       = C:\Program Files\Microsoft Visual Studio 8\Vc\bin
-DEFINE VS2005_DLL       = C:\Program Files\Microsoft Visual Studio 8\Common7\IDE;DEF(VS2005_BIN)
-DEFINE VS2005_BINX64    = C:\Program Files\Microsoft Visual Studio 8\Vc\bin\x86_amd64
-DEFINE VS2005_BIN64     = C:\Program Files\Microsoft Visual Studio 8\Vc\bin\x86_ia64
+DEFINE VS2005_BIN       = ENV(VS2005_PATH)\Vc\bin
+DEFINE VS2005_DLL       = ENV(VS2005_PATH)\Common7\IDE;DEF(VS2005_BIN)
+DEFINE VS2005_BINX64    = DEF(VS2005_BIN)\x86_amd64
+DEFINE VS2005_BIN64     = DEF(VS2005_BIN)\x86_ia64
 
 DEFINE VS2005x86_BIN    = C:\Program Files (x86)\Microsoft Visual Studio 8\Vc\bin
 DEFINE VS2005x86_DLL    = C:\Program Files (x86)\Microsoft Visual Studio 8\Common7\IDE;DEF(VS2005x86_BIN)
@@ -28,8 +28,8 @@
 DEFINE VS2005x86_BINX64 = DEF(VS2005x86_BIN)\x86_amd64
 DEFINE VS2005x86_BIN64  = DEF(VS2005x86_BIN)\x86_ia64
 
-DEFINE VS2008_BIN      = C:\Program Files\Microsoft Visual Studio 9.0\Vc\bin
-DEFINE VS2008_DLL      = C:\Program Files\Microsoft Visual Studio 9.0\Common7\IDE;DEF(VS2008_BIN)
+DEFINE VS2008_BIN      = ENV(VS2008_PATH)\Vc\bin
+DEFINE VS2008_DLL      = ENV(VS2008_PATH)\Common7\IDE;DEF(VS2008_BIN)
 DEFINE VS2008_BINX64   = DEF(VS2008_BIN)\x86_amd64
 DEFINE VS2008_BIN64    = DEF(VS2008_BIN)\x86_ia64
 
@@ -38,8 +38,8 @@
 DEFINE VS2008x86_BINX64 = DEF(VS2008x86_BIN)\x86_amd64
 DEFINE VS2008x86_BIN64  = DEF(VS2008x86_BIN)\x86_ia64
 
-DEFINE VS2010_BIN      = C:\Program Files\Microsoft Visual Studio 10.0\Vc\bin
-DEFINE VS2010_DLL      = C:\Program Files\Microsoft Visual Studio 10.0\Common7\IDE;DEF(VS2010_BIN)
+DEFINE VS2010_BIN      = ENV(VS2010_PATH)\Vc\bin
+DEFINE VS2010_DLL      = ENV(VS2010_PATH)\Common7\IDE;DEF(VS2010_BIN)
 DEFINE VS2010_BINX64   = DEF(VS2010_BIN)\x86_amd64
 DEFINE VS2010_BIN64    = DEF(VS2010_BIN)\x86_ia64
 
@@ -48,8 +48,8 @@
 DEFINE VS2010x86_BINX64 = DEF(VS2010x86_BIN)\x86_amd64
 DEFINE VS2010x86_BIN64  = DEF(VS2010x86_BIN)\x86_ia64
 
-DEFINE VS2012_BIN      = C:\Program Files\Microsoft Visual Studio 11.0\Vc\bin
-DEFINE VS2012_DLL      = C:\Program Files\Microsoft Visual Studio 11.0\Common7\IDE;DEF(VS2012_BIN)
+DEFINE VS2012_BIN      = ENV(VS2012_PATH)\Vc\bin
+DEFINE VS2012_DLL      = ENV(VS2012_PATH)\Common7\IDE;DEF(VS2012_BIN)
 DEFINE VS2012_BINX64   = DEF(VS2012_BIN)\x86_amd64
 
 DEFINE VS2012x86_BIN    = C:\Program Files (x86)\Microsoft Visual Studio 11.0\Vc\bin
@@ -674,8 +674,8 @@
 *_VS2003_IA32_DLINK_PATH            = DEF(VS2003_BIN)\link.exe
 *_VS2003_IA32_APP_PATH              = DEF(VS2003_BIN)\cl.exe
 *_VS2003_IA32_PP_PATH               = DEF(VS2003_BIN)\cl.exe
-*_VS2003_IA32_ASM_PATH              = DEF(VS2003_BIN)\ml.exe
-*_VS2003_IA32_ASM16_PATH            = DEF(VS2003_BIN)\ml.exe
+*_VS2003_IA32_ASM_PATH              = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
+*_VS2003_IA32_ASM16_PATH            = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2003_IA32_ASLCC_PATH            = DEF(VS2003_BIN)\cl.exe
 *_VS2003_IA32_ASLPP_PATH            = DEF(VS2003_BIN)\cl.exe
 *_VS2003_IA32_ASLDLINK_PATH         = DEF(VS2003_BIN)\link.exe
@@ -748,7 +748,7 @@
 *_VS2003xASL_IA32_DLINK_PATH            = DEF(VS2003_BIN)\link.exe
 *_VS2003xASL_IA32_APP_PATH              = DEF(VS2003_BIN)\cl.exe
 *_VS2003xASL_IA32_PP_PATH               = DEF(VS2003_BIN)\cl.exe
-*_VS2003xASL_IA32_ASM_PATH              = DEF(VS2003_BIN)\ml.exe
+*_VS2003xASL_IA32_ASM_PATH              = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2003xASL_IA32_ASLCC_PATH            = DEF(VS2003_BIN)\cl.exe
 *_VS2003xASL_IA32_ASLPP_PATH            = DEF(VS2003_BIN)\cl.exe
 *_VS2003xASL_IA32_ASLDLINK_PATH         = DEF(VS2003_BIN)\link.exe
@@ -810,7 +810,7 @@
 *_VS2005_*_PP_FLAGS               = /nologo /E /TC /FIAutoGen.h
 *_VS2005_*_VFRPP_FLAGS            = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2005_*_ASM16_PATH             = DEF(VS2005_BIN)\ml.exe
+*_VS2005_*_ASM16_PATH             = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -834,7 +834,7 @@
 *_VS2005_IA32_DLINK_PATH          = DEF(VS2005_BIN)\link.exe
 *_VS2005_IA32_APP_PATH            = DEF(VS2005_BIN)\cl.exe
 *_VS2005_IA32_PP_PATH             = DEF(VS2005_BIN)\cl.exe
-*_VS2005_IA32_ASM_PATH            = DEF(VS2005_BIN)\ml.exe
+*_VS2005_IA32_ASM_PATH            = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2005_IA32_ASLCC_PATH          = DEF(VS2005_BIN)\cl.exe
 *_VS2005_IA32_ASLPP_PATH          = DEF(VS2005_BIN)\cl.exe
 *_VS2005_IA32_ASLDLINK_PATH       = DEF(VS2005_BIN)\link.exe
@@ -861,7 +861,7 @@
 *_VS2005_X64_PP_PATH       = DEF(VS2005_BINX64)\cl.exe
 *_VS2005_X64_APP_PATH      = DEF(VS2005_BINX64)\cl.exe
 *_VS2005_X64_VFRPP_PATH    = DEF(VS2005_BINX64)\cl.exe
-*_VS2005_X64_ASM_PATH      = DEF(VS2005_BINX64)\ml64.exe
+*_VS2005_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2005_X64_SLINK_PATH    = DEF(VS2005_BINX64)\lib.exe
 *_VS2005_X64_DLINK_PATH    = DEF(VS2005_BINX64)\link.exe
 *_VS2005_X64_ASLCC_PATH    = DEF(VS2005_BINX64)\cl.exe
@@ -946,7 +946,7 @@
 *_VS2005xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2005xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2005xASL_*_ASM16_PATH      = DEF(VS2005_BIN)\ml.exe
+*_VS2005xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -970,7 +970,7 @@
 *_VS2005xASL_IA32_DLINK_PATH   = DEF(VS2005_BIN)\link.exe
 *_VS2005xASL_IA32_APP_PATH     = DEF(VS2005_BIN)\cl.exe
 *_VS2005xASL_IA32_PP_PATH      = DEF(VS2005_BIN)\cl.exe
-*_VS2005xASL_IA32_ASM_PATH     = DEF(VS2005_BIN)\ml.exe
+*_VS2005xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2005xASL_IA32_ASLCC_PATH    = DEF(VS2005_BIN)\cl.exe
 *_VS2005xASL_IA32_ASLPP_PATH    = DEF(VS2005_BIN)\cl.exe
 *_VS2005xASL_IA32_ASLDLINK_PATH = DEF(VS2005_BIN)\link.exe
@@ -997,7 +997,7 @@
 *_VS2005xASL_X64_PP_PATH       = DEF(VS2005_BINX64)\cl.exe
 *_VS2005xASL_X64_APP_PATH      = DEF(VS2005_BINX64)\cl.exe
 *_VS2005xASL_X64_VFRPP_PATH    = DEF(VS2005_BINX64)\cl.exe
-*_VS2005xASL_X64_ASM_PATH      = DEF(VS2005_BINX64)\ml64.exe
+*_VS2005xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2005xASL_X64_SLINK_PATH    = DEF(VS2005_BINX64)\lib.exe
 *_VS2005xASL_X64_DLINK_PATH    = DEF(VS2005_BINX64)\link.exe
 *_VS2005xASL_X64_ASLCC_PATH    = DEF(VS2005_BINX64)\cl.exe
@@ -1082,7 +1082,7 @@
 *_VS2005x86_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2005x86_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2005x86_*_ASM16_PATH      = DEF(VS2005x86_BIN)\ml.exe
+*_VS2005x86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1106,7 +1106,7 @@
 *_VS2005x86_IA32_DLINK_PATH   = DEF(VS2005x86_BIN)\link.exe
 *_VS2005x86_IA32_APP_PATH     = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86_IA32_PP_PATH      = DEF(VS2005x86_BIN)\cl.exe
-*_VS2005x86_IA32_ASM_PATH     = DEF(VS2005x86_BIN)\ml.exe
+*_VS2005x86_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2005x86_IA32_ASLCC_PATH    = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86_IA32_ASLPP_PATH    = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86_IA32_ASLDLINK_PATH = DEF(VS2005x86_BIN)\link.exe
@@ -1133,7 +1133,7 @@
 *_VS2005x86_X64_PP_PATH       = DEF(VS2005x86_BINX64)\cl.exe
 *_VS2005x86_X64_APP_PATH      = DEF(VS2005x86_BINX64)\cl.exe
 *_VS2005x86_X64_VFRPP_PATH    = DEF(VS2005x86_BINX64)\cl.exe
-*_VS2005x86_X64_ASM_PATH      = DEF(VS2005x86_BINX64)\ml64.exe
+*_VS2005x86_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2005x86_X64_SLINK_PATH    = DEF(VS2005x86_BINX64)\lib.exe
 *_VS2005x86_X64_DLINK_PATH    = DEF(VS2005x86_BINX64)\link.exe
 *_VS2005x86_X64_ASLCC_PATH    = DEF(VS2005x86_BINX64)\cl.exe
@@ -1218,7 +1218,7 @@
 *_VS2005x86xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2005x86xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2005x86xASL_*_ASM16_PATH      = DEF(VS2005x86_BIN)\ml.exe
+*_VS2005x86xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1242,7 +1242,7 @@
 *_VS2005x86xASL_IA32_DLINK_PATH   = DEF(VS2005x86_BIN)\link.exe
 *_VS2005x86xASL_IA32_APP_PATH     = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86xASL_IA32_PP_PATH      = DEF(VS2005x86_BIN)\cl.exe
-*_VS2005x86xASL_IA32_ASM_PATH     = DEF(VS2005x86_BIN)\ml.exe
+*_VS2005x86xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2005x86xASL_IA32_ASLCC_PATH    = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86xASL_IA32_ASLPP_PATH    = DEF(VS2005x86_BIN)\cl.exe
 *_VS2005x86xASL_IA32_ASLDLINK_PATH = DEF(VS2005x86_BIN)\link.exe
@@ -1269,7 +1269,7 @@
 *_VS2005x86xASL_X64_PP_PATH       = DEF(VS2005x86_BINX64)\cl.exe
 *_VS2005x86xASL_X64_APP_PATH      = DEF(VS2005x86_BINX64)\cl.exe
 *_VS2005x86xASL_X64_VFRPP_PATH    = DEF(VS2005x86_BINX64)\cl.exe
-*_VS2005x86xASL_X64_ASM_PATH      = DEF(VS2005x86_BINX64)\ml64.exe
+*_VS2005x86xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2005x86xASL_X64_SLINK_PATH    = DEF(VS2005x86_BINX64)\lib.exe
 *_VS2005x86xASL_X64_DLINK_PATH    = DEF(VS2005x86_BINX64)\link.exe
 *_VS2005x86xASL_X64_ASLCC_PATH    = DEF(VS2005x86_BINX64)\cl.exe
@@ -1355,7 +1355,7 @@
 *_VS2008_*_PP_FLAGS               = /nologo /E /TC /FIAutoGen.h
 *_VS2008_*_VFRPP_FLAGS            = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2008_*_ASM16_PATH             = DEF(VS2008_BIN)\ml.exe
+*_VS2008_*_ASM16_PATH             = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1379,7 +1379,7 @@
 *_VS2008_IA32_DLINK_PATH          = DEF(VS2008_BIN)\link.exe
 *_VS2008_IA32_APP_PATH            = DEF(VS2008_BIN)\cl.exe
 *_VS2008_IA32_PP_PATH             = DEF(VS2008_BIN)\cl.exe
-*_VS2008_IA32_ASM_PATH            = DEF(VS2008_BIN)\ml.exe
+*_VS2008_IA32_ASM_PATH            = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2008_IA32_ASLCC_PATH          = DEF(VS2008_BIN)\cl.exe
 *_VS2008_IA32_ASLPP_PATH          = DEF(VS2008_BIN)\cl.exe
 *_VS2008_IA32_ASLDLINK_PATH       = DEF(VS2008_BIN)\link.exe
@@ -1406,7 +1406,7 @@
 *_VS2008_X64_PP_PATH       = DEF(VS2008_BINX64)\cl.exe
 *_VS2008_X64_APP_PATH      = DEF(VS2008_BINX64)\cl.exe
 *_VS2008_X64_VFRPP_PATH    = DEF(VS2008_BINX64)\cl.exe
-*_VS2008_X64_ASM_PATH      = DEF(VS2008_BINX64)\ml64.exe
+*_VS2008_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2008_X64_SLINK_PATH    = DEF(VS2008_BINX64)\lib.exe
 *_VS2008_X64_DLINK_PATH    = DEF(VS2008_BINX64)\link.exe
 *_VS2008_X64_ASLCC_PATH    = DEF(VS2008_BINX64)\cl.exe
@@ -1493,7 +1493,7 @@
 *_VS2008xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2008xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2008xASL_*_ASM16_PATH      = DEF(VS2008_BIN)\ml.exe
+*_VS2008xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1517,7 +1517,7 @@
 *_VS2008xASL_IA32_DLINK_PATH   = DEF(VS2008_BIN)\link.exe
 *_VS2008xASL_IA32_APP_PATH     = DEF(VS2008_BIN)\cl.exe
 *_VS2008xASL_IA32_PP_PATH      = DEF(VS2008_BIN)\cl.exe
-*_VS2008xASL_IA32_ASM_PATH     = DEF(VS2008_BIN)\ml.exe
+*_VS2008xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2008xASL_IA32_ASLCC_PATH    = DEF(VS2008_BIN)\cl.exe
 *_VS2008xASL_IA32_ASLPP_PATH    = DEF(VS2008_BIN)\cl.exe
 *_VS2008xASL_IA32_ASLDLINK_PATH = DEF(VS2008_BIN)\link.exe
@@ -1544,7 +1544,7 @@
 *_VS2008xASL_X64_PP_PATH       = DEF(VS2008_BINX64)\cl.exe
 *_VS2008xASL_X64_APP_PATH      = DEF(VS2008_BINX64)\cl.exe
 *_VS2008xASL_X64_VFRPP_PATH    = DEF(VS2008_BINX64)\cl.exe
-*_VS2008xASL_X64_ASM_PATH      = DEF(VS2008_BINX64)\ml64.exe
+*_VS2008xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2008xASL_X64_SLINK_PATH    = DEF(VS2008_BINX64)\lib.exe
 *_VS2008xASL_X64_DLINK_PATH    = DEF(VS2008_BINX64)\link.exe
 *_VS2008xASL_X64_ASLCC_PATH    = DEF(VS2008_BINX64)\cl.exe
@@ -1629,7 +1629,7 @@
 *_VS2008x86_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2008x86_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2008x86_*_ASM16_PATH      = DEF(VS2008x86_BIN)\ml.exe
+*_VS2008x86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1656,7 +1656,7 @@
 *_VS2008x86_IA32_ASLDLINK_PATH= DEF(VS2008x86_BIN)\link.exe
 *_VS2008x86_IA32_APP_PATH     = DEF(VS2008x86_BIN)\cl.exe
 *_VS2008x86_IA32_PP_PATH      = DEF(VS2008x86_BIN)\cl.exe
-*_VS2008x86_IA32_ASM_PATH     = DEF(VS2008x86_BIN)\ml.exe
+*_VS2008x86_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2008x86_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2008x86_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -1682,7 +1682,7 @@
 *_VS2008x86_X64_VFRPP_PATH    = DEF(VS2008x86_BINX64)\cl.exe
 *_VS2008x86_X64_ASLCC_PATH    = DEF(VS2008x86_BINX64)\cl.exe
 *_VS2008x86_X64_ASLPP_PATH    = DEF(VS2008x86_BINX64)\cl.exe
-*_VS2008x86_X64_ASM_PATH      = DEF(VS2008x86_BINX64)\ml64.exe
+*_VS2008x86_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2008x86_X64_SLINK_PATH    = DEF(VS2008x86_BINX64)\lib.exe
 *_VS2008x86_X64_DLINK_PATH    = DEF(VS2008x86_BINX64)\link.exe
 *_VS2008x86_X64_ASLDLINK_PATH = DEF(VS2008x86_BINX64)\link.exe
@@ -1764,7 +1764,7 @@
 *_VS2008x86xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2008x86xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2008x86xASL_*_ASM16_PATH      = DEF(VS2008x86_BIN)\ml.exe
+*_VS2008x86xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1791,7 +1791,7 @@
 *_VS2008x86xASL_IA32_ASLDLINK_PATH= DEF(VS2008x86_BIN)\link.exe
 *_VS2008x86xASL_IA32_APP_PATH     = DEF(VS2008x86_BIN)\cl.exe
 *_VS2008x86xASL_IA32_PP_PATH      = DEF(VS2008x86_BIN)\cl.exe
-*_VS2008x86xASL_IA32_ASM_PATH     = DEF(VS2008x86_BIN)\ml.exe
+*_VS2008x86xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2008x86xASL_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2008x86xASL_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -1817,7 +1817,7 @@
 *_VS2008x86xASL_X64_VFRPP_PATH    = DEF(VS2008x86_BINX64)\cl.exe
 *_VS2008x86xASL_X64_ASLCC_PATH    = DEF(VS2008x86_BINX64)\cl.exe
 *_VS2008x86xASL_X64_ASLPP_PATH    = DEF(VS2008x86_BINX64)\cl.exe
-*_VS2008x86xASL_X64_ASM_PATH      = DEF(VS2008x86_BINX64)\ml64.exe
+*_VS2008x86xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2008x86xASL_X64_SLINK_PATH    = DEF(VS2008x86_BINX64)\lib.exe
 *_VS2008x86xASL_X64_DLINK_PATH    = DEF(VS2008x86_BINX64)\link.exe
 *_VS2008x86xASL_X64_ASLDLINK_PATH = DEF(VS2008x86_BINX64)\link.exe
@@ -1902,7 +1902,7 @@
 *_VS2010_*_PP_FLAGS               = /nologo /E /TC /FIAutoGen.h
 *_VS2010_*_VFRPP_FLAGS            = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2010_*_ASM16_PATH             = DEF(VS2010_BIN)\ml.exe
+*_VS2010_*_ASM16_PATH             = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -1926,7 +1926,7 @@
 *_VS2010_IA32_DLINK_PATH          = DEF(VS2010_BIN)\link.exe
 *_VS2010_IA32_APP_PATH            = DEF(VS2010_BIN)\cl.exe
 *_VS2010_IA32_PP_PATH             = DEF(VS2010_BIN)\cl.exe
-*_VS2010_IA32_ASM_PATH            = DEF(VS2010_BIN)\ml.exe
+*_VS2010_IA32_ASM_PATH            = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2010_IA32_ASLCC_PATH          = DEF(VS2010_BIN)\cl.exe
 *_VS2010_IA32_ASLPP_PATH          = DEF(VS2010_BIN)\cl.exe
 *_VS2010_IA32_ASLDLINK_PATH       = DEF(VS2010_BIN)\link.exe
@@ -1953,7 +1953,7 @@
 *_VS2010_X64_PP_PATH       = DEF(VS2010_BINX64)\cl.exe
 *_VS2010_X64_APP_PATH      = DEF(VS2010_BINX64)\cl.exe
 *_VS2010_X64_VFRPP_PATH    = DEF(VS2010_BINX64)\cl.exe
-*_VS2010_X64_ASM_PATH      = DEF(VS2010_BINX64)\ml64.exe
+*_VS2010_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2010_X64_SLINK_PATH    = DEF(VS2010_BINX64)\lib.exe
 *_VS2010_X64_DLINK_PATH    = DEF(VS2010_BINX64)\link.exe
 *_VS2010_X64_ASLCC_PATH    = DEF(VS2010_BINX64)\cl.exe
@@ -2040,7 +2040,7 @@
 *_VS2010xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2010xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2010xASL_*_ASM16_PATH      = DEF(VS2010_BIN)\ml.exe
+*_VS2010xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2064,7 +2064,7 @@
 *_VS2010xASL_IA32_DLINK_PATH   = DEF(VS2010_BIN)\link.exe
 *_VS2010xASL_IA32_APP_PATH     = DEF(VS2010_BIN)\cl.exe
 *_VS2010xASL_IA32_PP_PATH      = DEF(VS2010_BIN)\cl.exe
-*_VS2010xASL_IA32_ASM_PATH     = DEF(VS2010_BIN)\ml.exe
+*_VS2010xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2010xASL_IA32_ASLCC_PATH    = DEF(VS2010_BIN)\cl.exe
 *_VS2010xASL_IA32_ASLPP_PATH    = DEF(VS2010_BIN)\cl.exe
 *_VS2010xASL_IA32_ASLDLINK_PATH = DEF(VS2010_BIN)\link.exe
@@ -2091,7 +2091,7 @@
 *_VS2010xASL_X64_PP_PATH       = DEF(VS2010_BINX64)\cl.exe
 *_VS2010xASL_X64_APP_PATH      = DEF(VS2010_BINX64)\cl.exe
 *_VS2010xASL_X64_VFRPP_PATH    = DEF(VS2010_BINX64)\cl.exe
-*_VS2010xASL_X64_ASM_PATH      = DEF(VS2010_BINX64)\ml64.exe
+*_VS2010xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2010xASL_X64_SLINK_PATH    = DEF(VS2010_BINX64)\lib.exe
 *_VS2010xASL_X64_DLINK_PATH    = DEF(VS2010_BINX64)\link.exe
 *_VS2010xASL_X64_ASLCC_PATH    = DEF(VS2010_BINX64)\cl.exe
@@ -2176,7 +2176,7 @@
 *_VS2010x86_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2010x86_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2010x86_*_ASM16_PATH      = DEF(VS2010x86_BIN)\ml.exe
+*_VS2010x86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2203,7 +2203,7 @@
 *_VS2010x86_IA32_ASLDLINK_PATH= DEF(VS2010x86_BIN)\link.exe
 *_VS2010x86_IA32_APP_PATH     = DEF(VS2010x86_BIN)\cl.exe
 *_VS2010x86_IA32_PP_PATH      = DEF(VS2010x86_BIN)\cl.exe
-*_VS2010x86_IA32_ASM_PATH     = DEF(VS2010x86_BIN)\ml.exe
+*_VS2010x86_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2010x86_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2010x86_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -2229,7 +2229,7 @@
 *_VS2010x86_X64_VFRPP_PATH    = DEF(VS2010x86_BINX64)\cl.exe
 *_VS2010x86_X64_ASLCC_PATH    = DEF(VS2010x86_BINX64)\cl.exe
 *_VS2010x86_X64_ASLPP_PATH    = DEF(VS2010x86_BINX64)\cl.exe
-*_VS2010x86_X64_ASM_PATH      = DEF(VS2010x86_BINX64)\ml64.exe
+*_VS2010x86_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2010x86_X64_SLINK_PATH    = DEF(VS2010x86_BINX64)\lib.exe
 *_VS2010x86_X64_DLINK_PATH    = DEF(VS2010x86_BINX64)\link.exe
 *_VS2010x86_X64_ASLDLINK_PATH = DEF(VS2010x86_BINX64)\link.exe
@@ -2311,7 +2311,7 @@
 *_VS2010x86xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2010x86xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2010x86xASL_*_ASM16_PATH      = DEF(VS2010x86_BIN)\ml.exe
+*_VS2010x86xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2338,7 +2338,7 @@
 *_VS2010x86xASL_IA32_ASLDLINK_PATH= DEF(VS2010x86_BIN)\link.exe
 *_VS2010x86xASL_IA32_APP_PATH     = DEF(VS2010x86_BIN)\cl.exe
 *_VS2010x86xASL_IA32_PP_PATH      = DEF(VS2010x86_BIN)\cl.exe
-*_VS2010x86xASL_IA32_ASM_PATH     = DEF(VS2010x86_BIN)\ml.exe
+*_VS2010x86xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2010x86xASL_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2010x86xASL_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -2364,7 +2364,7 @@
 *_VS2010x86xASL_X64_VFRPP_PATH    = DEF(VS2010x86_BINX64)\cl.exe
 *_VS2010x86xASL_X64_ASLCC_PATH    = DEF(VS2010x86_BINX64)\cl.exe
 *_VS2010x86xASL_X64_ASLPP_PATH    = DEF(VS2010x86_BINX64)\cl.exe
-*_VS2010x86xASL_X64_ASM_PATH      = DEF(VS2010x86_BINX64)\ml64.exe
+*_VS2010x86xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2010x86xASL_X64_SLINK_PATH    = DEF(VS2010x86_BINX64)\lib.exe
 *_VS2010x86xASL_X64_DLINK_PATH    = DEF(VS2010x86_BINX64)\link.exe
 *_VS2010x86xASL_X64_ASLDLINK_PATH = DEF(VS2010x86_BINX64)\link.exe
@@ -2449,7 +2449,7 @@
 *_VS2012_*_PP_FLAGS               = /nologo /E /TC /FIAutoGen.h
 *_VS2012_*_VFRPP_FLAGS            = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2012_*_ASM16_PATH             = DEF(VS2012_BIN)\ml.exe
+*_VS2012_*_ASM16_PATH             = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2473,7 +2473,7 @@
 *_VS2012_IA32_DLINK_PATH          = DEF(VS2012_BIN)\link.exe
 *_VS2012_IA32_APP_PATH            = DEF(VS2012_BIN)\cl.exe
 *_VS2012_IA32_PP_PATH             = DEF(VS2012_BIN)\cl.exe
-*_VS2012_IA32_ASM_PATH            = DEF(VS2012_BIN)\ml.exe
+*_VS2012_IA32_ASM_PATH            = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2012_IA32_ASLCC_PATH          = DEF(VS2012_BIN)\cl.exe
 *_VS2012_IA32_ASLPP_PATH          = DEF(VS2012_BIN)\cl.exe
 *_VS2012_IA32_ASLDLINK_PATH       = DEF(VS2012_BIN)\link.exe
@@ -2500,7 +2500,7 @@
 *_VS2012_X64_PP_PATH       = DEF(VS2012_BINX64)\cl.exe
 *_VS2012_X64_APP_PATH      = DEF(VS2012_BINX64)\cl.exe
 *_VS2012_X64_VFRPP_PATH    = DEF(VS2012_BINX64)\cl.exe
-*_VS2012_X64_ASM_PATH      = DEF(VS2012_BINX64)\ml64.exe
+*_VS2012_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2012_X64_SLINK_PATH    = DEF(VS2012_BINX64)\lib.exe
 *_VS2012_X64_DLINK_PATH    = DEF(VS2012_BINX64)\link.exe
 *_VS2012_X64_ASLCC_PATH    = DEF(VS2012_BINX64)\cl.exe
@@ -2559,7 +2559,7 @@
 *_VS2012xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2012xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2012xASL_*_ASM16_PATH      = DEF(VS2012_BIN)\ml.exe
+*_VS2012xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2583,7 +2583,7 @@
 *_VS2012xASL_IA32_DLINK_PATH   = DEF(VS2012_BIN)\link.exe
 *_VS2012xASL_IA32_APP_PATH     = DEF(VS2012_BIN)\cl.exe
 *_VS2012xASL_IA32_PP_PATH      = DEF(VS2012_BIN)\cl.exe
-*_VS2012xASL_IA32_ASM_PATH     = DEF(VS2012_BIN)\ml.exe
+*_VS2012xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_VS2012xASL_IA32_ASLCC_PATH    = DEF(VS2012_BIN)\cl.exe
 *_VS2012xASL_IA32_ASLPP_PATH    = DEF(VS2012_BIN)\cl.exe
 *_VS2012xASL_IA32_ASLDLINK_PATH = DEF(VS2012_BIN)\link.exe
@@ -2610,7 +2610,7 @@
 *_VS2012xASL_X64_PP_PATH       = DEF(VS2012_BINX64)\cl.exe
 *_VS2012xASL_X64_APP_PATH      = DEF(VS2012_BINX64)\cl.exe
 *_VS2012xASL_X64_VFRPP_PATH    = DEF(VS2012_BINX64)\cl.exe
-*_VS2012xASL_X64_ASM_PATH      = DEF(VS2012_BINX64)\ml64.exe
+*_VS2012xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2012xASL_X64_SLINK_PATH    = DEF(VS2012_BINX64)\lib.exe
 *_VS2012xASL_X64_DLINK_PATH    = DEF(VS2012_BINX64)\link.exe
 *_VS2012xASL_X64_ASLCC_PATH    = DEF(VS2012_BINX64)\cl.exe
@@ -2667,7 +2667,7 @@
 *_VS2012x86_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2012x86_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2012x86_*_ASM16_PATH      = DEF(VS2012x86_BIN)\ml.exe
+*_VS2012x86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2694,7 +2694,7 @@
 *_VS2012x86_IA32_ASLDLINK_PATH= DEF(VS2012x86_BIN)\link.exe
 *_VS2012x86_IA32_APP_PATH     = DEF(VS2012x86_BIN)\cl.exe
 *_VS2012x86_IA32_PP_PATH      = DEF(VS2012x86_BIN)\cl.exe
-*_VS2012x86_IA32_ASM_PATH     = DEF(VS2012x86_BIN)\ml.exe
+*_VS2012x86_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2012x86_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2012x86_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -2720,7 +2720,7 @@
 *_VS2012x86_X64_VFRPP_PATH    = DEF(VS2012x86_BINX64)\cl.exe
 *_VS2012x86_X64_ASLCC_PATH    = DEF(VS2012x86_BINX64)\cl.exe
 *_VS2012x86_X64_ASLPP_PATH    = DEF(VS2012x86_BINX64)\cl.exe
-*_VS2012x86_X64_ASM_PATH      = DEF(VS2012x86_BINX64)\ml64.exe
+*_VS2012x86_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2012x86_X64_SLINK_PATH    = DEF(VS2012x86_BINX64)\lib.exe
 *_VS2012x86_X64_DLINK_PATH    = DEF(VS2012x86_BINX64)\link.exe
 *_VS2012x86_X64_ASLDLINK_PATH = DEF(VS2012x86_BINX64)\link.exe
@@ -2774,7 +2774,7 @@
 *_VS2012x86xASL_*_PP_FLAGS        = /nologo /E /TC /FIAutoGen.h
 *_VS2012x86xASL_*_VFRPP_FLAGS     = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_VS2012x86xASL_*_ASM16_PATH      = DEF(VS2012x86_BIN)\ml.exe
+*_VS2012x86xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2801,7 +2801,7 @@
 *_VS2012x86xASL_IA32_ASLDLINK_PATH= DEF(VS2012x86_BIN)\link.exe
 *_VS2012x86xASL_IA32_APP_PATH     = DEF(VS2012x86_BIN)\cl.exe
 *_VS2012x86xASL_IA32_PP_PATH      = DEF(VS2012x86_BIN)\cl.exe
-*_VS2012x86xASL_IA32_ASM_PATH     = DEF(VS2012x86_BIN)\ml.exe
+*_VS2012x86xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
       *_VS2012x86xASL_IA32_MAKE_FLAGS  = /nologo
   DEBUG_VS2012x86xASL_IA32_CC_FLAGS    = /nologo /c /WX /GS- /W4 /Gs32768 /D UNICODE /O1ib2 /GL /FIAutoGen.h /EHs-c- /GR- /GF /Gy /Zi /Gm
@@ -2827,7 +2827,7 @@
 *_VS2012x86xASL_X64_VFRPP_PATH    = DEF(VS2012x86_BINX64)\cl.exe
 *_VS2012x86xASL_X64_ASLCC_PATH    = DEF(VS2012x86_BINX64)\cl.exe
 *_VS2012x86xASL_X64_ASLPP_PATH    = DEF(VS2012x86_BINX64)\cl.exe
-*_VS2012x86xASL_X64_ASM_PATH      = DEF(VS2012x86_BINX64)\ml64.exe
+*_VS2012x86xASL_X64_ASM_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_VS2012x86xASL_X64_SLINK_PATH    = DEF(VS2012x86_BINX64)\lib.exe
 *_VS2012x86xASL_X64_DLINK_PATH    = DEF(VS2012x86_BINX64)\link.exe
 *_VS2012x86xASL_X64_ASLDLINK_PATH = DEF(VS2012x86_BINX64)\link.exe
@@ -2884,7 +2884,7 @@
 *_DDK3790_*_SLINK_FLAGS          = /nologo /LTCG
 *_DDK3790_*_VFRPP_FLAGS          = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_DDK3790_*_ASM16_PATH      = DEF(WINDDK_BIN32)\ml.exe
+*_DDK3790_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -2905,7 +2905,7 @@
 *_DDK3790_IA32_PP_PATH      = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790_IA32_VFRPP_PATH   = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790_IA32_APP_PATH     = DEF(WINDDK_BIN32)\cl.exe
-*_DDK3790_IA32_ASM_PATH     = DEF(WINDDK_BIN32)\ml.exe
+*_DDK3790_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_DDK3790_IA32_ASLCC_PATH     = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790_IA32_ASLPP_PATH     = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790_IA32_ASLDLINK_PATH  = DEF(WINDDK_BIN32)\link.exe
@@ -2931,7 +2931,7 @@
 *_DDK3790_X64_PP_PATH      = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790_X64_VFRPP_PATH   = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790_X64_APP_PATH     = DEF(WINDDK_BINX64)\cl.exe
-*_DDK3790_X64_ASM_PATH     = DEF(WINDDK_BINX64)\ml64.exe
+*_DDK3790_X64_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_DDK3790_X64_ASLCC_PATH     = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790_X64_ASLPP_PATH     = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790_X64_ASLDLINK_PATH  = DEF(WINDDK_BINX64)\link.exe
@@ -3009,7 +3009,7 @@
 *_DDK3790xASL_*_SLINK_FLAGS          = /nologo /LTCG
 *_DDK3790xASL_*_VFRPP_FLAGS          = /nologo /E /TC /DVFRCOMPILE /FI$(MODULE_NAME)StrDefs.h
 
-*_DDK3790xASL_*_ASM16_PATH      = DEF(WINDDK_BIN32)\ml.exe
+*_DDK3790xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -3030,7 +3030,7 @@
 *_DDK3790xASL_IA32_PP_PATH      = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790xASL_IA32_VFRPP_PATH   = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790xASL_IA32_APP_PATH     = DEF(WINDDK_BIN32)\cl.exe
-*_DDK3790xASL_IA32_ASM_PATH     = DEF(WINDDK_BIN32)\ml.exe
+*_DDK3790xASL_IA32_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_DDK3790xASL_IA32_ASLCC_PATH     = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790xASL_IA32_ASLPP_PATH     = DEF(WINDDK_BIN32)\cl.exe
 *_DDK3790xASL_IA32_ASLDLINK_PATH  = DEF(WINDDK_BIN32)\link.exe
@@ -3056,7 +3056,7 @@
 *_DDK3790xASL_X64_PP_PATH      = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790xASL_X64_VFRPP_PATH   = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790xASL_X64_APP_PATH     = DEF(WINDDK_BINX64)\cl.exe
-*_DDK3790xASL_X64_ASM_PATH     = DEF(WINDDK_BINX64)\ml64.exe
+*_DDK3790xASL_X64_ASM_PATH     = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_DDK3790xASL_X64_ASLCC_PATH     = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790xASL_X64_ASLPP_PATH     = DEF(WINDDK_BINX64)\cl.exe
 *_DDK3790xASL_X64_ASLDLINK_PATH  = DEF(WINDDK_BINX64)\link.exe
@@ -4085,7 +4085,7 @@
 *_ICC_*_APP_FLAGS                        = /nologo /E /TC
 *_ICC_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICC_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICC_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4108,7 +4108,7 @@
 *_ICC_IA32_PP_PATH                  = DEF(ICC_BIN32)\icl.exe
 *_ICC_IA32_VFRPP_PATH               = DEF(ICC_BIN32)\icl.exe
 *_ICC_IA32_APP_PATH                 = DEF(ICC_BIN32)\icl.exe
-*_ICC_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICC_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICC_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICC_IA32_ASLCC_PATH               = DEF(ICC_BIN32)\icl.exe
 *_ICC_IA32_ASLPP_PATH               = DEF(ICC_BIN32)\icl.exe
@@ -4138,7 +4138,7 @@
 *_ICC_X64_PP_PATH                   = DEF(ICC_BINX64)\icl.exe
 *_ICC_X64_VFRPP_PATH                = DEF(ICC_BINX64)\icl.exe
 *_ICC_X64_APP_PATH                  = DEF(ICC_BINX64)\icl.exe
-*_ICC_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICC_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICC_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICC_X64_ASLCC_PATH                = DEF(ICC_BINX64)\icl.exe
 *_ICC_X64_ASLPP_PATH                = DEF(ICC_BINX64)\icl.exe
@@ -4229,7 +4229,7 @@
 *_ICCxASL_*_APP_FLAGS                        = /nologo /E /TC
 *_ICCxASL_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICCxASL_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICCxASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4252,7 +4252,7 @@
 *_ICCxASL_IA32_PP_PATH                  = DEF(ICC_BIN32)\icl.exe
 *_ICCxASL_IA32_VFRPP_PATH               = DEF(ICC_BIN32)\icl.exe
 *_ICCxASL_IA32_APP_PATH                 = DEF(ICC_BIN32)\icl.exe
-*_ICCxASL_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICCxASL_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICCxASL_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICCxASL_IA32_ASLCC_PATH               = DEF(ICC_BIN32)\icl.exe
 *_ICCxASL_IA32_ASLPP_PATH               = DEF(ICC_BIN32)\icl.exe
@@ -4282,7 +4282,7 @@
 *_ICCxASL_X64_PP_PATH                   = DEF(ICC_BINX64)\icl.exe
 *_ICCxASL_X64_VFRPP_PATH                = DEF(ICC_BINX64)\icl.exe
 *_ICCxASL_X64_APP_PATH                  = DEF(ICC_BINX64)\icl.exe
-*_ICCxASL_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICCxASL_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICCxASL_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICCxASL_X64_ASLCC_PATH                = DEF(ICC_BINX64)\icl.exe
 *_ICCxASL_X64_ASLPP_PATH                = DEF(ICC_BINX64)\icl.exe
@@ -4374,7 +4374,7 @@
 *_ICCx86_*_APP_FLAGS                        = /nologo /E /TC
 *_ICCx86_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICCx86_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICCx86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4397,7 +4397,7 @@
 *_ICCx86_IA32_PP_PATH                  = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86_IA32_VFRPP_PATH               = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86_IA32_APP_PATH                 = DEF(ICC_BIN32x86)\icl.exe
-*_ICCx86_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICCx86_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICCx86_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICCx86_IA32_ASLCC_PATH               = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86_IA32_ASLPP_PATH               = DEF(ICC_BIN32x86)\icl.exe
@@ -4427,7 +4427,7 @@
 *_ICCx86_X64_PP_PATH                   = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86_X64_VFRPP_PATH                = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86_X64_APP_PATH                  = DEF(ICC_BINX64x86)\icl.exe
-*_ICCx86_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICCx86_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICCx86_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICCx86_X64_ASLCC_PATH                = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86_X64_ASLPP_PATH                = DEF(ICC_BINX64x86)\icl.exe
@@ -4519,7 +4519,7 @@
 *_ICCx86xASL_*_APP_FLAGS                        = /nologo /E /TC
 *_ICCx86xASL_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICCx86ASL_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICCx86ASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4542,7 +4542,7 @@
 *_ICCx86xASL_IA32_PP_PATH                  = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86xASL_IA32_VFRPP_PATH               = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86xASL_IA32_APP_PATH                 = DEF(ICC_BIN32x86)\icl.exe
-*_ICCx86xASL_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICCx86xASL_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICCx86xASL_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICCx86xASL_IA32_ASLCC_PATH               = DEF(ICC_BIN32x86)\icl.exe
 *_ICCx86xASL_IA32_ASLPP_PATH               = DEF(ICC_BIN32x86)\icl.exe
@@ -4572,7 +4572,7 @@
 *_ICCx86xASL_X64_PP_PATH                   = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86xASL_X64_VFRPP_PATH                = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86xASL_X64_APP_PATH                  = DEF(ICC_BINX64x86)\icl.exe
-*_ICCx86xASL_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICCx86xASL_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICCx86xASL_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICCx86xASL_X64_ASLCC_PATH                = DEF(ICC_BINX64x86)\icl.exe
 *_ICCx86xASL_X64_ASLPP_PATH                = DEF(ICC_BINX64x86)\icl.exe
@@ -4665,7 +4665,7 @@
 *_ICC11_*_APP_FLAGS                        = /nologo /E /TC
 *_ICC11_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICC11_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICC11_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4688,7 +4688,7 @@
 *_ICC11_IA32_PP_PATH                  = DEF(ICC11_BIN32)\icl.exe
 *_ICC11_IA32_VFRPP_PATH               = DEF(ICC11_BIN32)\icl.exe
 *_ICC11_IA32_APP_PATH                 = DEF(ICC11_BIN32)\icl.exe
-*_ICC11_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICC11_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICC11_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICC11_IA32_ASLCC_PATH               = DEF(ICC11_BIN32)\icl.exe
 *_ICC11_IA32_ASLPP_PATH               = DEF(ICC11_BIN32)\icl.exe
@@ -4717,7 +4717,7 @@
 *_ICC11_X64_PP_PATH                   = DEF(ICC11_BINX64)\icl.exe
 *_ICC11_X64_VFRPP_PATH                = DEF(ICC11_BINX64)\icl.exe
 *_ICC11_X64_APP_PATH                  = DEF(ICC11_BINX64)\icl.exe
-*_ICC11_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICC11_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICC11_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICC11_X64_ASLCC_PATH                = DEF(ICC11_BINX64)\icl.exe
 *_ICC11_X64_ASLPP_PATH                = DEF(ICC11_BINX64)\icl.exe
@@ -4808,7 +4808,7 @@
 *_ICC11xASL_*_APP_FLAGS                        = /nologo /E /TC
 *_ICC11xASL_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICC11xASL_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICC11xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4831,7 +4831,7 @@
 *_ICC11xASL_IA32_PP_PATH                  = DEF(ICC11_BIN32)\icl.exe
 *_ICC11xASL_IA32_VFRPP_PATH               = DEF(ICC11_BIN32)\icl.exe
 *_ICC11xASL_IA32_APP_PATH                 = DEF(ICC11_BIN32)\icl.exe
-*_ICC11xASL_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICC11xASL_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICC11xASL_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICC11xASL_IA32_ASLCC_PATH               = DEF(ICC11_BIN32)\icl.exe
 *_ICC11xASL_IA32_ASLPP_PATH               = DEF(ICC11_BIN32)\icl.exe
@@ -4860,7 +4860,7 @@
 *_ICC11xASL_X64_PP_PATH                   = DEF(ICC11_BINX64)\icl.exe
 *_ICC11xASL_X64_VFRPP_PATH                = DEF(ICC11_BINX64)\icl.exe
 *_ICC11xASL_X64_APP_PATH                  = DEF(ICC11_BINX64)\icl.exe
-*_ICC11xASL_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICC11xASL_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICC11xASL_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICC11xASL_X64_ASLCC_PATH                = DEF(ICC11_BINX64)\icl.exe
 *_ICC11xASL_X64_ASLPP_PATH                = DEF(ICC11_BINX64)\icl.exe
@@ -4953,7 +4953,7 @@
 *_ICC11x86_*_APP_FLAGS                        = /nologo /E /TC
 *_ICC11x86_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICC11x86_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICC11x86_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -4976,7 +4976,7 @@
 *_ICC11x86_IA32_PP_PATH                  = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86_IA32_VFRPP_PATH               = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86_IA32_APP_PATH                 = DEF(ICC11_BIN32x86)\icl.exe
-*_ICC11x86_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICC11x86_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICC11x86_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICC11x86_IA32_ASLCC_PATH               = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86_IA32_ASLPP_PATH               = DEF(ICC11_BIN32x86)\icl.exe
@@ -5004,7 +5004,7 @@
 *_ICC11x86_X64_PP_PATH                   = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86_X64_VFRPP_PATH                = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86_X64_APP_PATH                  = DEF(ICC11_BINX64x86)\icl.exe
-*_ICC11x86_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICC11x86_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICC11x86_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICC11x86_X64_ASLCC_PATH                = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86_X64_ASLPP_PATH                = DEF(ICC11_BINX64x86)\icl.exe
@@ -5095,7 +5095,7 @@
 *_ICC11x86xASL_*_APP_FLAGS                        = /nologo /E /TC
 *_ICC11x86xASL_*_PP_FLAGS                         = /nologo /E /TC /FIAutoGen.h
 
-*_ICC11x86xASL_*_ASM16_PATH      = DEF(MS_VS_BIN)\ml.exe
+*_ICC11x86xASL_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # ASL definitions
@@ -5118,7 +5118,7 @@
 *_ICC11x86xASL_IA32_PP_PATH                  = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86xASL_IA32_VFRPP_PATH               = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86xASL_IA32_APP_PATH                 = DEF(ICC11_BIN32x86)\icl.exe
-*_ICC11x86xASL_IA32_ASM_PATH                 = DEF(MS_VS_BIN)\ml.exe
+*_ICC11x86xASL_IA32_ASM_PATH                 = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_ICC11x86xASL_IA32_ASM_DLL                  = DEF(MS_VS_DLL)
 *_ICC11x86xASL_IA32_ASLCC_PATH               = DEF(ICC11_BIN32x86)\icl.exe
 *_ICC11x86xASL_IA32_ASLPP_PATH               = DEF(ICC11_BIN32x86)\icl.exe
@@ -5147,7 +5147,7 @@
 *_ICC11x86xASL_X64_PP_PATH                   = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86xASL_X64_VFRPP_PATH                = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86xASL_X64_APP_PATH                  = DEF(ICC11_BINX64x86)\icl.exe
-*_ICC11x86xASL_X64_ASM_PATH                  = DEF(WINDDK_BINX64)\ml64.exe
+*_ICC11x86xASL_X64_ASM_PATH                  = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_ICC11x86xASL_X64_ASM_DLL                   = DEF(MS_VS_DLL)
 *_ICC11x86xASL_X64_ASLCC_PATH                = DEF(ICC11_BINX64x86)\icl.exe
 *_ICC11x86xASL_X64_ASLPP_PATH                = DEF(ICC11_BINX64x86)\icl.exe
@@ -5245,7 +5245,7 @@
 *_MYTOOLS_*_PP_FLAGS                     = /nologo /E /TC /FIAutoGen.h
 *_MYTOOLS_*_SLINK_FLAGS                  = /nologo /LTCG
 
-*_MYTOOLS_*_ASM16_PATH      = DEF(VS2008_BIN)\ml.exe
+*_MYTOOLS_*_ASM16_PATH      = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 
 ##################
 # IA32 definitions
@@ -5259,7 +5259,7 @@
 *_MYTOOLS_IA32_PP_PATH              = DEF(VS2008_BIN)\cl.exe
 *_MYTOOLS_IA32_VFRPP_PATH           = DEF(VS2008_BIN)\cl.exe
 *_MYTOOLS_IA32_APP_PATH             = DEF(VS2008_BIN)\cl.exe
-*_MYTOOLS_IA32_ASM_PATH             = DEF(VS2008_BIN)\ml.exe
+*_MYTOOLS_IA32_ASM_PATH             = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p
 *_MYTOOLS_IA32_ASLCC_PATH           = DEF(VS2008_BIN)\cl.exe
 *_MYTOOLS_IA32_ASLPP_PATH           = DEF(VS2008_BIN)\cl.exe
 *_MYTOOLS_IA32_ASLDLINK_PATH        = DEF(VS2008_BIN)\link.exe
@@ -5287,7 +5287,7 @@
 *_MYTOOLS_X64_PP_PATH               = DEF(VS2008_BINX64)\cl.exe
 *_MYTOOLS_X64_VFRPP_PATH            = DEF(VS2008_BINX64)\cl.exe
 *_MYTOOLS_X64_APP_PATH              = DEF(VS2008_BINX64)\cl.exe
-*_MYTOOLS_X64_ASM_PATH              = DEF(VS2008_BINX64)\ml64.exe
+*_MYTOOLS_X64_ASM_PATH              = ENV(WORKSPACE)\BaseTools\Bin\Win32\JWASM.EXE -10p -win64
 *_MYTOOLS_X64_ASLCC_PATH            = DEF(VS2008_BINX64)\cl.exe
 *_MYTOOLS_X64_ASLPP_PATH            = DEF(VS2008_BINX64)\cl.exe
 *_MYTOOLS_X64_ASLDLINK_PATH         = DEF(VS2008_BINX64)\link.exe
Index: build-windows.bat
===================================================================
--- build-windows.bat	(revision 0)
+++ build-windows.bat	(working copy)
@@ -0,0 +1,230 @@
+@echo off
+rem ----------------project settings-------------------
+set PROJECT_NAME=DuetPkg
+rem # NOOPT   : good for source level, but slower execution
+rem # DEBUG   : fast execution, but register variables makes source level debug harder
+rem # RELEASE : fast execution, no source level debug and no serial out
+set BUILDTARGET=DEBUG
+
+rem # 0x80000000  Errors only 
+rem # 0x80000004  Errors and some module load messages
+rem # 0x8000004F  Lots of messages
+rem # 0x800000CF  Lots of messages, with dependency logging
+set DEBUG_PRINT_LEVEL=0x8000004F
+
+rem # compiler selection
+set TOOL_CHAIN_TAG=VS2008
+set TARGETARCH=X64
+
+rem -------------------memory map----------------------
+rem
+rem   000000-001000                     real mode vectors, used by BiosVideoThunkDxe
+rem   007000-007BFF                     stack (early real mode execution)
+rem   007C00-007DFF                     bootsect.asm
+rem   010000-0101FF                     temp disk read buffer used by Start64.asm
+rem   020000-021FFF IMAGE_BUFFER_LOW    first FAT cluster of EFILDR16 (16 sectors per cluster)
+rem   021800-021FFE IMAGE_BUFFER_LOW    stack (st16_64.asm)
+rem   020000-02FFFF IMAGE_BUFFER_LOW    EFILDR16 (first 64KB)
+rem   050000-05FFFF X64_STACK_TOP       initial x64 stack
+rem   070000-07FFFF FAKE_NVRAM_BASE     UEFI NVRAM emulation memory
+rem   080000-085FFF X64_PAGE_TABLE_BASE temp page tables for 64-bit enable
+rem   100000-xxxxxx IMAGE_BUFFER_HIGH   EFILDR16 (complete)
+
+rem --------these two values are hard-coded in the coreboot payload code--------
+set /a IMAGE_BUFFER_HIGH   = 0x100000
+set /a START64CODESIZE     = 0x2000
+rem ----------------------------------------------------------------------------
+set /a IMAGE_BUFFER_LOW    = 0x20000
+set /a X64_STACK_TOP       = 0x60000
+set /a FAKE_NVRAM_BASE     = 0x70000
+set /a FAKE_NVRAM_SIZE     = 0x10000
+set /a X64_PAGE_TABLE_BASE = 0x80000
+set /a X64_PAGE_TABLE_SIZE = 0x6000
+set /a EFI64CODESIZE       = 0x3000
+set /a EFI64CODESTART      = %IMAGE_BUFFER_HIGH% + %START64CODESIZE%
+
+rem # optional: display logo during UEFI startup
+set LOGO_FILE=DuetPkg/Coreboot_full_web.bmp
+
+rem --------------end project settings-----------------
+
+rem Make sure current dir is that of this batch file.
+cd  /d %~dp0
+
+rem # Locate the UEFI workspace directory.
+set WORKSPACE=
+set WORKSPACE_SEARCH0=%cd%
+set WORKSPACE_SEARCH1=%cd%\edk2
+if exist %WORKSPACE_SEARCH1%\MdePkg\ set WORKSPACE=%WORKSPACE_SEARCH1%
+if exist %WORKSPACE_SEARCH0%\MdePkg\ set WORKSPACE=%WORKSPACE_SEARCH0%
+if "%WORKSPACE%" == "" (
+    echo --- unable to locate uefi workspace directory
+    pause
+    exit /b 1
+    )
+
+set BASE_TOOLS_PATH=%WORKSPACE%\BaseTools
+set BUILD_DIR=%WORKSPACE%\build
+set IMAGE_DIR=%BUILD_DIR%\images
+set BASETOOLS_DIR=%BASE_TOOLS_PATH%\Bin\Win32
+set OUTPUT_DIR=%BUILD_DIR%\%PROJECT_NAME%\%BUILDTARGET%_%TOOL_CHAIN_TAG%
+set OUTPUT_DIR_PROC=%OUTPUT_DIR%\%TARGETARCH%
+set BOOTSECTOR_BIN_DIR=%OUTPUT_DIR_PROC%\DuetPkg\BootSector\BootSector\Output
+set PROJECT_DIR=%WORKSPACE%\%PROJECT_NAME%
+set PLATFORMFILE=%PROJECT_DIR%\DuetPkg%TARGETARCH%.dsc
+
+rem ---------------------------------------------------
+rem # arguments for the build command
+set /a THREADS=NUMBER_OF_PROCESSORS*2
+set _BUILD_OPTIONS_=^
+-p %PLATFORMFILE% ^
+-b %BUILDTARGET% ^
+-t %TOOL_CHAIN_TAG% ^
+-a %TARGETARCH% ^
+-D DEBUG_PRINT_LEVEL=%DEBUG_PRINT_LEVEL% ^
+-D BUILD_NEW_SHELL ^
+-D LOGO_FILE=%LOGO_FILE% ^
+-D X64_STACK_TOP=%X64_STACK_TOP% ^
+-D FAKE_NVRAM_BASE=%FAKE_NVRAM_BASE% ^
+-D FAKE_NVRAM_SIZE=%FAKE_NVRAM_SIZE% ^
+-n %THREADS%
+
+rem # create the B command (build with output sent to a text file)
+echo @call rombuild.bat ^> buildlog-windows.txt 2^>^&1 > b.bat
+echo @if errorlevel 1 echo ===== BUILD FAILED =====>> b.bat
+
+if "%TOOL_CHAIN_TAG%" == "VS2003" goto LOCATE_VS2003
+if "%TOOL_CHAIN_TAG%" == "VS2005" goto LOCATE_VS2005
+if "%TOOL_CHAIN_TAG%" == "VS2008" goto LOCATE_VS2008
+if "%TOOL_CHAIN_TAG%" == "VS2010" goto LOCATE_VS2010
+if "%TOOL_CHAIN_TAG%" == "VS2012" goto LOCATE_VS2012
+goto VS_NOT_FOUND
+
+:LOCATE_VS2003
+if "%VS71COMNTOOLS%" == "" goto VS_NOT_FOUND
+for %%? in ("%VS71COMNTOOLS%..\..") do set VS2003_PATH=%%~f?
+set VS_PATH=%VS2003_PATH%
+goto VS_FOUND
+
+:LOCATE_VS2005
+if "%VS80COMNTOOLS%" == "" goto VS_NOT_FOUND
+for %%? in ("%VS80COMNTOOLS%..\..") do set VS2005_PATH=%%~f?
+set VS_PATH=%VS2005_PATH%
+goto VS_FOUND
+
+:LOCATE_VS2008
+if "%VS90COMNTOOLS%" == "" goto VS_NOT_FOUND
+for %%? in ("%VS90COMNTOOLS%..\..") do set VS2008_PATH=%%~f?
+set VS_PATH=%VS2008_PATH%
+goto VS_FOUND
+
+:LOCATE_VS2010
+if "%VS100COMNTOOLS%" == "" goto VS_NOT_FOUND
+for %%? in ("%VS100COMNTOOLS%..\..") do set VS2010_PATH=%%~f?
+set VS_PATH=%VS2010_PATH%
+goto VS_FOUND
+
+:LOCATE_VS2012
+if "%VS110COMNTOOLS%" == "" goto VS_NOT_FOUND
+for %%? in ("%VS110COMNTOOLS%..\..") do set VS2012_PATH=%%~f?
+set VS_PATH=%VS2012_PATH%
+goto VS_FOUND
+
+:VS_NOT_FOUND
+if "%VS_PATH%" == "" (
+   echo.
+   echo =========================================
+   echo ===  MS %TOOL_CHAIN_TAG% build tools not found  ===
+   echo =========================================
+   echo.
+   pause
+   exit
+   )
+echo.
+
+:VS_FOUND
+REM # do a test compile to sanity check the build tool setup
+setlocal
+cd /d %temp%
+echo int main (void) {return 0;} > testapp.c
+set path=%VS_PATH%\VC\bin;%VS_PATH%\Common7\IDE
+cl.exe -c -nologo testapp.c >nul 2>nul
+IF %ERRORLEVEL% NEQ 0 (
+   echo.
+   echo --------------------------------------------------------
+   echo - UEFI compile tools found, but a test compile failed. -
+   echo - Check your tool setup.                               -
+   echo --------------------------------------------------------
+   echo.
+   pause
+   exit
+   )
+del testapp.* > nul
+endlocal
+
+rem # create the CLEAN command
+echo @setlocal                                                    > clean.bat
+echo @rd  %BUILD_DIR%                           /s/q ^>nul 2^>nul >> clean.bat
+echo @rd  %WORKSPACE%\Conf\.cache               /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\Conf\tools_def.txt        /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\Conf\build_rule.txt       /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\Conf\target.txt           /s/q ^>nul 2^>nul >> clean.bat
+echo @del *.pyc                                 /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\BaseTools\*.o             /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\BaseTools\*.d             /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\BaseTools\*.a             /s/q ^>nul 2^>nul >> clean.bat
+echo @del %WORKSPACE%\BaseTools\Source\C\bin\*. /s/q ^>nul 2^>nul >> clean.bat
+echo @endlocal                                                    >> clean.bat
+
+rem # create the ROMBUILD command
+echo @echo off                                                                                                       > rombuild.bat
+echo rem (generated file, edit build-windows.bat instead)                                                           >> rombuild.bat
+echo setlocal                                                                                                       >> rombuild.bat
+echo echo using compile tools from %UEFI_COMPILE_TOOLS%                                                             >> rombuild.bat
+echo rem # delete the rom image so that none is present after a build fail                                          >> rombuild.bat
+echo del %OUTPUT_DIR%\FV\*.Fv 2^>nul                                                                                >> rombuild.bat
+echo cd %~dp0                                                                                                       >> rombuild.bat
+echo call %BASE_TOOLS_PATH%\toolsetup.bat ^> nul                                                                    >> rombuild.bat
+echo @echo off                                                                                                      >> rombuild.bat
+echo build.exe %_BUILD_OPTIONS_%                                                                                    >> rombuild.bat
+echo if errorlevel 1 GOTO buildfail                                                                                 >> rombuild.bat
+echo IF not exist %OUTPUT_DIR%\FV\DUETEFIMAINFV.Fv GOTO buildfail                                                   >> rombuild.bat
+echo cd %%OUTPUT_DIR%%\FV                                                                                           >> rombuild.bat
+echo echo Compressing %%OUTPUT_DIR%%\FV\DUETEFIMAINFV.Fv...                                                         >> rombuild.bat
+echo %%BASETOOLS_DIR%%\LzmaCompress -e -o FWVOLUME.z DUETEFIMAINFV.Fv                                               >> rombuild.bat
+echo IF not %%errorlevel%% == 0 GOTO buildfail                                                                      >> rombuild.bat
+echo echo Compressing %%OUTPUT_DIR_PROC%%\DxeCore.efi ...                                                           >> rombuild.bat
+echo %%BASETOOLS_DIR%%\LzmaCompress -e -o DxeMain.z %%OUTPUT_DIR_PROC%%\DxeCore.efi                                 >> rombuild.bat
+echo IF not %%errorlevel%% == 0 GOTO buildfail                                                                      >> rombuild.bat
+echo echo Compressing %%OUTPUT_DIR_PROC%%\DxeIpl.efi ...                                                            >> rombuild.bat
+echo %%BASETOOLS_DIR%%\LzmaCompress -e -o DxeIpl.z %%OUTPUT_DIR_PROC%%\DxeIpl.efi                                   >> rombuild.bat
+echo IF not %%errorlevel%% == 0 GOTO buildfail                                                                      >> rombuild.bat
+echo echo Generate Loader Image Efildr64...                                                                         >> rombuild.bat
+echo %%BASETOOLS_DIR%%\GenFw.exe -r --rebase 0x10000 %%OUTPUT_DIR_PROC%%\EfiLoader.efi                              >> rombuild.bat
+echo %%BASETOOLS_DIR%%\EfiLdrImage.exe -o Efildr64 %%OUTPUT_DIR_PROC%%\EfiLoader.efi DxeIpl.z DxeMain.z FWVOLUME.z  >> rombuild.bat
+echo IF not %%errorlevel%% == 0 GOTO buildfail                                                                      >> rombuild.bat
+echo @copy /b %BOOTSECTOR_BIN_DIR%\St16_64.com+%BOOTSECTOR_BIN_DIR%\Efi64.com2+Efildr64 EFILDR16                    >> rombuild.bat
+echo md %%IMAGE_DIR%% 2^> nul                                                                                       >> rombuild.bat
+echo copy EFILDR16 %%IMAGE_DIR%% /y                                                                                 >> rombuild.bat
+echo echo Creating FAT disk image...                                                                                >> rombuild.bat
+echo cd %BOOTSECTOR_BIN_DIR%                                                                                        >> rombuild.bat
+echo BuildFatImage mbr=Mbr.com bs=Bs16.com ldr=%%IMAGE_DIR%%\EFILDR16 out=%%IMAGE_DIR%%\FAT16.img                   >> rombuild.bat
+echo if errorlevel 1 GOTO buildfail                                                                                 >> rombuild.bat
+echo cd %WORKSPACE%                                                                                                 >> rombuild.bat
+echo ECHO ===== BUILD SUCCESS =====                                                                                 >> rombuild.bat
+echo     exit /b 0                                                                                                  >> rombuild.bat
+echo :buildfail                                                                                                     >> rombuild.bat
+echo     ECHO ===== BUILD FAILED =====                                                                              >> rombuild.bat
+echo endlocal                                                                                                       >> rombuild.bat
+echo     exit /b 1                                                                                                  >> rombuild.bat
+
+@echo off
+
+if "%1" == "batch" rombuild.bat
+if "%1" == "" (
+   start "Duet build %cd%" %comspec% /k^
+   "echo compile tools   %UEFI_COMPILE_TOOLS%&&echo BASE_TOOLS_PATH %BASE_TOOLS_PATH%&&echo WORKSPACE       %WORKSPACE%&&echo PLATFORMFILE    %PLATFORMFILE%&&echo VS_PATH         %VS_PATH%&&echo.&&echo === build   ROMBUILD&&echo === clean   CLEAN"
+   ) else (
+   echo === unrecognized option: %1
+   pause
+)
