Index: DuetPkg/EfiLdr/EfiLoader.c
===================================================================
--- DuetPkg/EfiLdr/EfiLoader.c	(revision 1)
+++ DuetPkg/EfiLdr/EfiLoader.c	(working copy)
@@ -24,6 +24,9 @@
 #include "PeLoader.h"
 #include "LzmaDecompress.h"
 
+#define EFILDR_LOAD_ADDRESS        IMAGE_BUFFER_HIGH
+#define EFILDR_HEADER_ADDRESS      (EFILDR_LOAD_ADDRESS + START64CODESIZE + EFI64CODESIZE)
+
 VOID
 SystemHang (
   CHAR8        *Message
@@ -30,7 +33,7 @@
   )
 {
   PrintString (
-    "%s## FATAL ERROR ##: Fail to load DUET images! System hang!\n",
+    "%a## FATAL ERROR ##: Fail to load DUET images! System hang!\n",
     Message
     );
   CpuDeadLoop();
