Index: DuetPkg/PciBusNoEnumerationDxe/PciDeviceSupport.c
===================================================================
--- DuetPkg/PciBusNoEnumerationDxe/PciDeviceSupport.c	(revision 1)
+++ DuetPkg/PciBusNoEnumerationDxe/PciDeviceSupport.c	(working copy)
@@ -314,7 +314,7 @@
              );
   if (!EFI_ERROR (Status)) {
     PciIoDevice->IsPciExp = TRUE;
-    DEBUG ((EFI_D_ERROR, "PciExp - %x (B-%x, D-%x, F-%x)\n", PciIoDevice->IsPciExp, PciIoDevice->BusNumber, PciIoDevice->DeviceNumber, PciIoDevice->FunctionNumber));
+    DEBUG ((EFI_D_INFO, "PciExp - %x (B-%x, D-%x, F-%x)\n", PciIoDevice->IsPciExp, PciIoDevice->BusNumber, PciIoDevice->DeviceNumber, PciIoDevice->FunctionNumber));
   }
   
   //
