Index: PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcRtc.h
===================================================================
--- PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcRtc.h	(revision 1)
+++ PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcRtc.h	(working copy)
@@ -78,7 +78,12 @@
 //
 #define RTC_INIT_REGISTER_A 0x26
 #define RTC_INIT_REGISTER_B 0x02
-#define RTC_INIT_REGISTER_D 0x0
+// PcRtc.c wants to see register D bit 7 (VRT) high almost immediately
+// after writing the below value, which clears it with the default 
+// UEFI value of zero. The AMD SB700 updates this bit only once per
+// 4-1020ms (1020ms default). This causes function RtcWaitToUpdate
+// to return an error. Preset VRT to 1 to avoid this.
+#define RTC_INIT_REGISTER_D 0x80
 
 #pragma pack(1)
 //
