Index: DuetPkg/SataControllerDxe/SataController.c
===================================================================
--- DuetPkg/SataControllerDxe/SataController.c	(revision 1)
+++ DuetPkg/SataControllerDxe/SataController.c	(working copy)
@@ -411,7 +411,8 @@
                   EFI_OPEN_PROTOCOL_BY_DRIVER
                   );
   if (EFI_ERROR (Status)) {
-    DEBUG ((EFI_D_ERROR, "SataControllerStart error return status = %r\n", Status));
+    if (Status != EFI_ALREADY_STARTED)
+      DEBUG ((EFI_D_ERROR, "SataControllerStart error return status = %r\n", Status));
     return Status;
   }
 
