Index: MdeModulePkg/Bus/Ata/AtaAtapiPassThru/IdeMode.c
===================================================================
--- MdeModulePkg/Bus/Ata/AtaAtapiPassThru/IdeMode.c	(revision 1)
+++ MdeModulePkg/Bus/Ata/AtaAtapiPassThru/IdeMode.c	(working copy)
@@ -2598,9 +2598,9 @@
     //
     IdeWritePortB (PciIo, IdeRegisters->CmdOrStatus, ATA_CMD_EXEC_DRIVE_DIAG);
 
-    Status = WaitForBSYClear (PciIo, IdeRegisters, 350000000);
+    Status = WaitForBSYClear (PciIo, IdeRegisters, 3500000);
     if (EFI_ERROR (Status)) {
-      DEBUG((EFI_D_ERROR, "New detecting method: Send Execute Diagnostic Command: WaitForBSYClear: Status: %d\n", Status));
+      DEBUG((EFI_D_INFO, "New detecting method: Send Execute Diagnostic Command: WaitForBSYClear: Status: %d\n", Status));
       continue;
     }
 
