Index: MdeModulePkg/Universal/Console/TerminalDxe/TerminalConOut.c
===================================================================
--- MdeModulePkg/Universal/Console/TerminalDxe/TerminalConOut.c	(revision 1)
+++ MdeModulePkg/Universal/Console/TerminalDxe/TerminalConOut.c	(working copy)
@@ -13,6 +13,7 @@
 **/
 
 #include "Terminal.h"
+#include <Library/PrintLib.h>
 
 //
 // This list is used to define the valid extend chars.
@@ -81,7 +82,7 @@
 CHAR16 mSetModeString[]            = { ESC, '[', '=', '3', 'h', 0 };
 CHAR16 mSetAttributeString[]       = { ESC, '[', '0', 'm', ESC, '[', '4', '0', 'm', ESC, '[', '4', '0', 'm', 0 };
 CHAR16 mClearScreenString[]        = { ESC, '[', '2', 'J', 0 };
-CHAR16 mSetCursorPositionString[]  = { ESC, '[', '0', '0', ';', '0', '0', 'H', 0 };
+CHAR16 mSetCursorPositionString[12];
 
 //
 // Body of the ConOut functions
@@ -755,12 +756,9 @@
   //
   // control sequence to move the cursor
   //
-  mSetCursorPositionString[ROW_OFFSET + 0]    = (CHAR16) ('0' + ((Row + 1) / 10));
-  mSetCursorPositionString[ROW_OFFSET + 1]    = (CHAR16) ('0' + ((Row + 1) % 10));
-  mSetCursorPositionString[COLUMN_OFFSET + 0] = (CHAR16) ('0' + ((Column + 1) / 10));
-  mSetCursorPositionString[COLUMN_OFFSET + 1] = (CHAR16) ('0' + ((Column + 1) % 10));
+  UnicodeSPrint (mSetCursorPositionString, sizeof mSetCursorPositionString, L"\x1B[%d;%dH", (int)Row + 1, (int)Column + 1);
 
-  TerminalDevice->OutputEscChar               = TRUE;
+  TerminalDevice->OutputEscChar = TRUE;
   Status = This->OutputString (This, mSetCursorPositionString);
   TerminalDevice->OutputEscChar = FALSE;
 
