Index: DuetPkg/DuetPkg.fdf
===================================================================
--- DuetPkg/DuetPkg.fdf	(revision 13)
+++ DuetPkg/DuetPkg.fdf	(working copy)
@@ -73,7 +73,7 @@
 
 INF  PcAtChipsetPkg/8259InterruptControllerDxe/8259.inf
 INF  DuetPkg/AcpiResetDxe/Reset.inf
-INF  DuetPkg/LegacyMetronome/Metronome.inf
+INF  MdeModulePkg/Universal/Metronome/Metronome.inf
 INF  PcAtChipsetPkg/8254TimerDxe/8254Timer.inf
 INF  PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcatRealTimeClockRuntimeDxe.inf
 INF  DuetPkg/PciRootBridgeNoEnumerationDxe/PciRootBridgeNoEnumeration.inf
Index: DuetPkg/DuetPkgX64.dsc
===================================================================
--- DuetPkg/DuetPkgX64.dsc	(revision 13)
+++ DuetPkg/DuetPkgX64.dsc	(working copy)
@@ -216,7 +216,7 @@
   UefiCpuPkg/CpuDxe/CpuDxe.inf
   PcAtChipsetPkg/8259InterruptControllerDxe/8259.inf
   DuetPkg/AcpiResetDxe/Reset.inf
-  DuetPkg/LegacyMetronome/Metronome.inf
+  MdeModulePkg/Universal/Metronome/Metronome.inf
 
   PcAtChipsetPkg/PcatRealTimeClockRuntimeDxe/PcatRealTimeClockRuntimeDxe.inf
   PcAtChipsetPkg/8254TimerDxe/8254Timer.inf
Index: DuetPkg/LegacyMetronome/Metronome.c
===================================================================
--- DuetPkg/LegacyMetronome/Metronome.c	(revision 1)
+++ DuetPkg/LegacyMetronome/Metronome.c	(working copy)
@@ -1,125 +0,0 @@
-/*++
-
-Copyright (c) 2005, Intel Corporation. All rights reserved.<BR>
-This program and the accompanying materials                          
-are licensed and made available under the terms and conditions of the BSD License         
-which accompanies this distribution.  The full text of the license may be found at        
-http://opensource.org/licenses/bsd-license.php                                            
-                                                                                          
-THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,                     
-WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.             
-
-Module Name:
-  LegacyMetronome.c
-
-Abstract:
-
-  This contains the installation function for the driver.
-
---*/
-
-#include "Metronome.h"
-
-//
-// Handle for the Metronome Architectural Protocol instance produced by this driver
-//
-EFI_HANDLE                  mMetronomeHandle = NULL;
-
-//
-// The Metronome Architectural Protocol instance produced by this driver
-//
-EFI_METRONOME_ARCH_PROTOCOL mMetronome = {
-  WaitForTick,
-  TICK_PERIOD
-};
-
-//
-// Worker Functions
-//
-EFI_STATUS
-EFIAPI
-WaitForTick (
-  IN EFI_METRONOME_ARCH_PROTOCOL  *This,
-  IN UINT32                       TickNumber
-  )
-/*++
-
-Routine Description:
-
-  Waits for the TickNumber of ticks from a known platform time source.
-
-Arguments:
-
-  This                Pointer to the protocol instance.
-
-Returns: 
-
-  EFI_SUCCESS         If number of ticks occurred.
-  EFI_NOT_FOUND       Could not locate CPU IO protocol
-
---*/
-// TODO:    TickNumber - add argument and description to function comment
-{
-  //
-  // Wait for TickNumber toggles of the Refresh bit
-  //
-  for (; TickNumber != 0x00; TickNumber--) {
-    while ((IoRead8(REFRESH_PORT) & REFRESH_ON) == REFRESH_ON);
-    while ((IoRead8(REFRESH_PORT) & REFRESH_ON) == REFRESH_OFF);
-  }
-
-  return EFI_SUCCESS;
-}
-
-EFI_STATUS
-EFIAPI
-InstallMetronome (
-  IN EFI_HANDLE        ImageHandle,
-  IN EFI_SYSTEM_TABLE  *SystemTable
-  )
-/*++
-
-Routine Description:
-  
-  Install the LegacyMetronome driver.  Loads a Metronome Arch Protocol based
-  on the Port 61 timer.
-
-Arguments:
-
-  (Standard EFI Image entry - EFI_IMAGE_ENTRY_POINT)
-
-Returns:
-
-  EFI_SUCCESS - Metronome Architectural Protocol Installed
-
---*/
-// TODO:    ImageHandle - add argument and description to function comment
-// TODO:    SystemTable - add argument and description to function comment
-{
-  EFI_STATUS  Status;
-
-  //
-  // Make sure the Metronome Architectural Protocol is not already installed in the system
-  //
-  ASSERT_PROTOCOL_ALREADY_INSTALLED (NULL, &gEfiMetronomeArchProtocolGuid);
-
-  //
-  // Program port 61 timer 1 as refresh timer. We could use ACPI timer in the
-  // future.
-  //
-  IoWrite8 (TIMER1_CONTROL_PORT, LOAD_COUNTER1_LSB);
-  IoWrite8 (TIMER1_COUNT_PORT, COUNTER1_COUNT);
-
-  //
-  // Install on a new handle
-  //
-  Status = gBS->InstallMultipleProtocolInterfaces (
-                  &mMetronomeHandle,
-                  &gEfiMetronomeArchProtocolGuid,
-                  &mMetronome,
-                  NULL
-                  );
-  ASSERT_EFI_ERROR (Status);
-
-  return Status;
-}
Index: DuetPkg/LegacyMetronome/Metronome.h
===================================================================
--- DuetPkg/LegacyMetronome/Metronome.h	(revision 1)
+++ DuetPkg/LegacyMetronome/Metronome.h	(working copy)
@@ -1,75 +0,0 @@
-/*++
-
-Copyright (c) 2005, Intel Corporation. All rights reserved.<BR>
-This program and the accompanying materials                          
-are licensed and made available under the terms and conditions of the BSD License         
-which accompanies this distribution.  The full text of the license may be found at        
-http://opensource.org/licenses/bsd-license.php                                            
-                                                                                          
-THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,                     
-WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.             
-
-Module Name:
-  LegacyMetronome.h
-
-Abstract:
-
-  Driver implementing the EFI 2.0 metronome protocol using the legacy PORT 61 
-  timer.
-
---*/
-
-#ifndef _LEGACY_METRONOME_H
-#define _LEGACY_METRONOME_H
-
-//
-// Statements that include other files
-//
-#include <PiDxe.h>
-
-#include <Protocol/Metronome.h>
-
-#include <Library/DebugLib.h>
-#include <Library/UefiBootServicesTableLib.h>
-#include <Library/IoLib.h>
-
-//
-// Private definitions
-//
-#define TICK_PERIOD         300
-#define REFRESH_PORT        0x61
-#define REFRESH_ON          0x10
-#define REFRESH_OFF         0x00
-#define TIMER1_CONTROL_PORT 0x43
-#define TIMER1_COUNT_PORT   0x41
-#define LOAD_COUNTER1_LSB   0x54
-#define COUNTER1_COUNT      0x12
-
-//
-// Function Prototypes
-//
-EFI_STATUS
-EFIAPI
-WaitForTick (
-  IN EFI_METRONOME_ARCH_PROTOCOL  *This,
-  IN UINT32                       TickNumber
-  )
-/*++
-
-Routine Description:
-
-  TODO: Add function description
-
-Arguments:
-
-  This        - TODO: add argument description
-  TickNumber  - TODO: add argument description
-
-Returns:
-
-  TODO: add return values
-
---*/
-;
-
-#endif
Index: DuetPkg/LegacyMetronome/Metronome.inf
===================================================================
--- DuetPkg/LegacyMetronome/Metronome.inf	(revision 1)
+++ DuetPkg/LegacyMetronome/Metronome.inf	(working copy)
@@ -1,43 +0,0 @@
-## @file
-# 
-# Copyright (c) 2005 - 2010, Intel Corporation. All rights reserved.<BR>
-# This program and the accompanying materials                          
-# are licensed and made available under the terms and conditions of the BSD License         
-# which accompanies this distribution.  The full text of the license may be found at        
-# http://opensource.org/licenses/bsd-license.php                                            
-#                                                                                           
-# THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,                     
-# WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.             
-#
-#  Module Name:
-#
-#  Abstract:
-#
-##
-
-[Defines]
-  INF_VERSION                    = 0x00010005
-  BASE_NAME                      = Metronome
-  FILE_GUID                      = 07A9330A-F347-11d4-9A49-0090273FC14D
-  MODULE_TYPE                    = DXE_DRIVER
-  VERSION_STRING                 = 1.0
-  ENTRY_POINT                    = InstallMetronome
-
-[Packages]
-  MdePkg/MdePkg.dec
-
-[LibraryClasses]
-  UefiDriverEntryPoint
-  DebugLib
-  UefiBootServicesTableLib
-  IoLib
-  
-[Sources]
-  Metronome.c
-  Metronome.h
-
-[Protocols]
-  gEfiMetronomeArchProtocolGuid
-
-[Depex]
-  TRUE
